### ISSUE TITLE
cue: structural cycles not detected

### ISSUE THREAD

**Issue by cueckoo** [cue-lang/cue#42]
_Originally opened by @fabxc in https://github.com/cuelang/cue/issues/42_

cue seems to detect what the spec describes as reference cycles. But for structural cycles, it runs into a stack overflow:

```
goroutine 1 [running]:
runtime.heapBitsSetType(0xc04ab64070, 0x8, 0x8, 0xc5bcc0)
        /usr/lib/google-golang/src/runtime/mbitmap.go:938 +0xa55 fp=0xc028000350 sp=0xc028000348 pc=0x52cea5
runtime.mallocgc(0x8, 0xc5bcc0, 0x1, 0x0)
        /usr/lib/google-golang/src/runtime/malloc.go:969 +0x50e fp=0xc0280003f0 sp=0xc028000350 pc=0x522c3e
runtime.newobject(0xc5bcc0, 0x0)
        /usr/lib/google-golang/src/runtime/malloc.go:1068 +0x38 fp=0xc028000420 sp=0xc0280003f0 pc=0x523298
cuelang.org/go/cue.(*structLit).expandFields(0xc00025c000, 0xc01907de60, 0xc01907de60)
        /usr/local/google/home/freinartz/repos/src/cuelang.org/go/cue/value.go:646 +0x42 fp=0xc028000658 sp=0xc028000420 pc=0x7d88c2
cuelang.org/go/cue.Value.structValOpts(0xc000240b90, 0xc04ab63a00, 0xc01907de60, 0x101000000, 0xc04ab63a00, 0xc00025c000, 0xd98c20, 0xc00025c000, 0xd96780)
        /usr/local/google/home/freinartz/repos/src/cuelang.org/go/cue/types.go:813 +0xcd fp=0xc028000788 sp=0xc028000658 pc=0x7d239d
cuelang.org/go/cue.Value.structVal(...)
        /usr/local/google/home/freinartz/repos/src/cuelang.org/go/cue/types.go:799
cuelang.org/go/cue.Value.Walk(0xc000240b90, 0xc04ab63a00, 0xc047fff870, 0x0)
        /usr/local/google/home/freinartz/repos/src/cuelang.org/go/cue/types.go:1126 +0xe9 fp=0xc028000890 sp=0xc028000788 pc=0x7d4649
cuelang.org/go/cue.Value.Walk(0xc000240b90, 0xc04ab639c0, 0xc047fff870, 0x0)
        /usr/local/google/home/freinartz/repos/src/cuelang.org/go/cue/types.go:1129 +0x161 fp=0xc028000998 sp=0xc028000890 pc=0x7d46c1
cuelang.org/go/cue.Value.Walk(0xc000240b90, 0xc04ab63980, 0xc047fff870, 0x0)
        /usr/local/google/home/freinartz/repos/src/cuelang.org/go/cue/types.go:1129 +0x161 fp=0xc028000aa0 sp=0xc028000998 pc=0x7d46c1
cuelang.org/go/cue.Value.Walk(0xc000240b90, 0xc04ab63940, 0xc047fff870, 0x0)
...
```

**Comment by cueckoo** [comment]
_Original reply by @mpvl in https://github.com/cuelang/cue/issues/42#issuecomment-489139265_

Yes, there is an cryptic open issue for that: #29: implement occurs check.

The reason why this is not done yet is that the exact implementation has many consequences. For instance, do we want to support primitive recursive functions, the Ackerman function, or none of the above?

For the engine/formalism itself it is fine, btw. It can reason over infinite structure. It just cannot output it, so it makes sense to limit it.

**Comment by cueckoo** [comment]
_Original reply by @mpvl in https://github.com/cuelang/cue/issues/42#issuecomment-489147622_

Although it is a duplicate, I'll leave it in as the title is probably more discoverable. :)

**Comment by cueckoo** [comment]
_Original reply by @mpvl in https://github.com/cuelang/cue/issues/42#issuecomment-663976141_

Fixed in v0.3.0-alpha1. See https://github.com/cuelang/cue/blob/master/cue/testdata/cycle/structural.txtar for test cases.
